src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -751,6 +752,65 @@ class RandomDateGenerator : public Generator<bsoncxx::types::b_date> {
     const UniqueGenerator<int64_t> _maxGen;
 };
 
+class CycleGenerator : public Appendable {
+
+public:
+    CycleGenerator(const Node& node,
+                    GeneratorArgs generatorArgs,
+                    std::map<std::string, Parser<UniqueAppendable>> parsers)
+        : CycleGenerator(
+            extract(node, "of", "^Cycle"),
+            generatorArgs,
+            parsers,
+            extract(node, "number", "^Cycle").to<int64_t>()) {}
+
+    CycleGenerator(const Node& node,
+                GeneratorArgs generatorArgs,
+                std::map<std::string, Parser<UniqueAppendable>> parsers,
+                int64_t size)
+    : _cacheSize{size},
+      _cache{generateCache(node, generatorArgs, parsers, _cacheSize)},
+      _index{0} {}
+
+    void append(const std::string& key, bsoncxx::builder::basic::document& builder) override {
+        auto cacheView = _cache.view();
+        builder.append(bsoncxx::builder::basic::kvp(key, cacheView[_index].get_value()));
+        updateIndex();
+    }
+    void append(bsoncxx::builder::basic::array& builder) override {
+        auto cacheView = _cache.view();
+        builder.append(cacheView[_index].get_value());
+        updateIndex();
+    }
+
+private:
+    static bsoncxx::array::value generateCache(const Node& node,
+                    GeneratorArgs generatorArgs,
+                    std::map<std::string, Parser<UniqueAppendable>> parsers,
+                    int64_t size) {
+        bsoncxx::builder::basic::array builder{};
+        auto valueGen = valueGenerator<false, UniqueAppendable>(node,
```

Comment:
The ValueGenerator goes out of scope after initialization time, yes? It looks like it does -- I went back to specifically check. There's no good reason to keep them around.

Reply:
Yep it should go out of scope once the object is initialized